scheduling/taskmetadata.go

```diff
@@ -0,0 +1,550 @@
+			return "", e
+		}
+
+		return string(serializedMetadata), nil
+	}
+
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: metadata.StartTime}
+	e = m.storer.Update(taskRunMetadataKey.String(), updateStatus)
+	return e
+}
+
+func (m *TaskMetadataManager) SetRunEndTime(runID TaskRunID, taskID TaskID, time time.Time) fferr.GRPCError {
+	if taskID <= 0 {
+		errMessage := fmt.Errorf("task id has to be greater than 0")
+		err := fferr.NewInvalidArgumentError(errMessage)
+		err.AddDetail("task_id", string(taskID))
+		return err
+	}
+	if time.IsZero() {
+		errMessage := fmt.Errorf("end time cannot be zero")
+		err := fferr.NewInvalidArgumentError(errMessage)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		err.AddDetail("end_time", time.String())
+		return err
+	}
+
+	metadata, e := m.GetRunByID(taskID, runID)
+	if e != nil {
+		e.AddDetail("task_id", string(taskID))
+		e.AddDetail("run_id", string(runID))
+		return e
+	}
+
+	updateEndTime := func(runMetadata string) (string, fferr.GRPCError) {
+		metadata := TaskRunMetadata{}
+		err := metadata.Unmarshal([]byte(runMetadata))
+		if err != nil {
+			err.AddDetail("task_id", string(taskID))
+			err.AddDetail("run_id", string(runID))
+			return "", err
+		}
+
+		if metadata.StartTime.After(time) {
+			err := fferr.NewInvalidArgumentError(fmt.Errorf("end time cannot be before start time"))
+			err.AddDetail("task_id", string(taskID))
+			err.AddDetail("run_id", string(runID))
+			err.AddDetail("start_time", metadata.StartTime.String())
+			err.AddDetail("end_time", time.String())
+			return "", err
+		}
+
+		metadata.EndTime = time
+		serializedMetadata, e := metadata.Marshal()
+		if e != nil {
+			return "", e
+		}
+
+		return string(serializedMetadata), nil
+	}
+
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: metadata.StartTime}
+	e = m.storer.Update(taskRunMetadataKey.String(), updateEndTime)
+	return e
+}
+
+func (m *TaskMetadataManager) AppendRunLog(runID TaskRunID, taskID TaskID, log string) fferr.GRPCError {
+	if runID <= 0 {
+		err := fferr.NewInvalidArgumentError(fmt.Errorf("run id has to be greater than 0"))
+		err.AddDetail("run_id", string(runID))
+		err.AddDetail("task_id", string(taskID))
+		return err
+	}
+	if taskID <= 0 {
+		err := fferr.NewInvalidArgumentError(fmt.Errorf("task id has to be greater than 0"))
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return err
+	}
+	if log == "" {
+		err := fferr.NewInvalidArgumentError(fmt.Errorf("log cannot be empty"))
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return err
+	}
+
+	metadata, e := m.GetRunByID(taskID, runID)
+	if e != nil {
+		return e
+	}
+
+	updateLog := func(runMetadata string) (string, fferr.GRPCError) {
+		metadata := TaskRunMetadata{}
+		err := metadata.Unmarshal([]byte(runMetadata))
+		if err != nil {
+			return "", err
```

Comment:
is this supposed to have a structured error?

Reply:
they will return fferr types